src/jit/morph.cpp

```diff
@@ -6703,7 +6703,12 @@ GenTree* Compiler::fgMorphField(GenTree* tree, MorphAddrContext* mac)
                 else
                 {
 #if CONSERVATIVE_NULL_CHECK_BYREF_CREATION
-                    addExplicitNullCheck = (mac->m_kind == MACK_Addr && (mac->m_totalOffset + fldOffset > 0));
+                    // In R2R mode the field offset may be changed when the code is loaded.
+                    // So we can't rely on a zero offset here to suppress the null check.
+                    //
+                    // See GitHub issue #16454.
+                    addExplicitNullCheck =
+                        (mac->m_kind == MACK_Addr) && ((mac->m_totalOffset + fldOffset > 0) || opts.IsReadyToRun());
```

Comment:
This is only needed for `CORINFO_FIELD_INSTANCE_WITH_BASE` fields. I think you can check for it here using `gtField.gtFieldLookup.addr != nullptr`.  Most of the fields in R2R mode have fixed offsets.

Reply:
Thanks, updated.